doc/clojure.1
.\"to render: groff -Tascii -man doc/clojure.1 > clojure.man"
.TH CLOJURE 1 "2019 June 6"
.SH NAME
clj, clojure \- runners to launch Clojure programs

.SH SYNOPSIS

.B clj
[\fBclj-opts\fR] [\fB-Aaliases\fR]
.br
.B clojure
[\fBclj-opts\fR] \fB-X\fR[\fBaliases\fR] \fBmy/fn?\fR [\fBkpath v\fR ...] \fBkv-map?\fR
.br
.B clojure
[\fBclj-opts\fR] \fB-T\fR[\fBname|aliases\fR] \fBmy/fn\fR [\fBkpath v\fR ...] \fBkv-map?\fR
.br
.B clojure
[\fBclj-opts\fR] \fB-M\fR[\fBaliases\fR] [\fBinit-opts\fR] [\fBmain-opts\fR] [\fBargs\fR]

.SH DESCRIPTION

The
.B clojure
tool is a runner for Clojure programs.
.B clj
is a wrapper for
.B clojure
that uses rlwrap to provide a better interactive experience.
Users will typically run
.BR clj .

.SH EXAMPLES

You can launch a Clojure REPL by running
.BR clj .

See the Deps and CLI Guide below for more complex usage.

.SH OPTIONS

The
.B clj
and
.B clojure
scripts can take three types of options, as shown in the
synopsis.
.B clj-opts
are used to build the java-opts and
classpath.
.B init-opts
refer to Clojure code to execute.
.B main-opts
are options to clojure.main.

.SS clj-opts

.TP
.BI \-J opt
Pass opt through in java_opts, ex: -J-Xmx512m

.TP
.BI \-A aliases
Concatenated aliases for REPL execution, ex: -A:dev:mem

.TP
.BI \-X aliases\ \fIfn\fR " " \fIKPATH\fR " " \fIV\fR " "
Exec alias to invoke one or more functions that take a map,
with keypath/value overrides. Function must either be
supplied or be in :exec-fn argmap for alias

.TP
.BI \-M aliases
Exec clojure.main, either from opts in alias or command line

.TP
.BI \-P
Prepare - download libs, cache classpath without executing

.TP
.BI \-Sdeps\  EDN
Deps data or file to use as the last deps file to be merged

.TP
.BI \-Spath
Compute classpath and echo to stdout only

.TP
.BI \-Stree
Print dependency tree

.TP
.BI \-Scp\  CP
Do NOT compute or cache classpath, use this one instead

.TP
.BI \-Srepro
Ignore the ~/.clojure/deps.edn config file

.TP
.BI \-Sforce
Force recomputation of the classpath (don't use the cache)

.TP
.BI \-Sverbose
Print important path info to console

.TP
.BI \-Sthreads\  N
Set specific number of download threads

.TP
.BI \-Strace
Write a trace.edn file that traces deps expansion

.TP
.BI \-\-
Stop parsing dep options and pass remaining arguments to clojure.main

.TP
.BI \-\-version
Write version to stdout and exit

.TP
.BI \-version
Write version to stderr and exit

.SS init-opts

.TP
.BR \-i ", " \-\-init\  \fIpath\fR
Load a file or resource located at \fIpath\fR

.TP
.BR \-e ", " \-\-eval\  \fIstring\fR
Eval exprs in \fIstring\fR; print non-nil values

.TP
.BR \-\-report\  \fItarget\fR
Report uncaught exception to \fItarget\fR: "file" (default), "stderr", or "none", overrides System property clojure.main.report

.SS main-opts


.TP
.BR \-m ", " \-\-main\  \fIns-name \fR
Call the -main function from namespace w/args

.TP
.BR \-r ", " \-\-repl
Run a repl

.TP
.I path
Run a script from a file or resource located at \fIpath\fR

.TP
.B \-
Run a script from standard input

.TP
.BR \-h ", " \-? ",  " \-\-help
Print this help message and exit

.SH CONFIGURATION

This script reads configuration from
.B deps.edn
files. It will check against
.B deps.edn
files in the system root location, the user
configuration directory (usually
.BR ~/.clojure ),
and the local directory. Files in each of these locations, if
they exist, are merged to form one combined configuration
file. The last specified configuration file's options
overwrites any earlier files (e.g. local configs take
priority).

.SH SEE ALSO

.IP https://clojure.org/guides/deps_and_cli
Deps and CLI Guide
.IP https://clojure.org/reference/deps_edn
deps.edn Reference
.IP https://clojure.org/reference/clojure_cli
Clojure CLI Reference
.IP https://clojure.org/reference/repl_and_main
REPL and Main Entrypoints
.IP https://clojure.org/releases/tools
Release history and changelog

.SH BUGS

File issues at
https://ask.clojure.org under category Libs / tools.deps.

.SH AUTHORS

clojure and clj are maintained by Alex Miller <alex.miller@cognitect.com>.
This man page was written by Elana Hashman <ehashman@debian.org>.

.SH COPYRIGHT

Copyright \(co Rich Hickey, Alex Miller and contributors.

Distributed under the Eclipse Public License 1.0, the same as
Clojure.


CHANGELOG.md
Changelog
===========

**Also see [tools.deps changelog](https://github.com/clojure/tools.deps.alpha/blob/master/CHANGELOG.md)**

clj has both stable and prerelease versions. Current and former stable build are listed in **bold** and are (or were) available from the default [brew formula](https://github.com/clojure/brew-install/). Other versions can be obtained using versioned formulas only.

Prerelease versions: none

**Current stable version:**

* **1.12.1.1550 on Jun 3, 2025**
  * **`-Sdescribe` - undocumented (prefer using tools.deps APIs instead)**
  * **`-Sverbose` - add explicit deps.edn references**

Older versions (previous stable builds in bold):

* **1.12.1.1543 on Jun 3, 2025**
  * **Update to Clojure 1.12.1**
* **1.12.1.1538 on Jun 2, 2025**
  * **Update to Clojure 1.12.1**
* **1.12.0.1530 on Mar 5, 2025**
  * **Update to tools.deps 0.23.1512**
* 1.12.0.1523 on Mar 5, 2025
  * Also accept file for -Sdeps, and check it for staleness
  * Update to tools.deps 0.22.1497
* **1.12.0.1517 on Feb 7, 2025**
  * **Update to tools.deps 0.22.1492**
* 1.12.0.1514 on Feb 7, 2025
  * Update to tools.deps 0.22.1484
* 1.12.0.1510 on Feb 6, 2025
  * Update to tools.deps 0.22.1480
* 1.12.0.1506 on Feb 6, 2025
  * Update to tools.deps 0.22.1476
* **1.12.0.1501 on Jan 27, 2025**
  * **Update to tools.deps 0.21.1471**
* **1.12.0.1495 on Dec 31, 2024**
  * **Echo args with -M deprecation warning for clarity**
  * **Update to tools.deps 0.21.1467**
* **1.12.0.1488** on Nov 21, 2024
  * **Tweak message when -X or -T function namespace not found**
  * **Update to latest tools.tools**
  * **Update to latest tools.deps**
* **1.12.0.1479**
  * **Default to Clojure 1.12.0**
  * **Update to tools.deps 0.21.1449 and tools.deps.cli**
* **1.11.4.1474**
  * **Default to Clojure 1.11.4**
  * **Update to tools.deps 0.20.1440**
* **1.11.3.1463**
  * **Update list of :deps programs in help**
  * **Update to latest tools.deps.cli, tools.deps, and tools.build**
* **1.11.3.1456 on Apr 24, 2024**
  * **Update to latest tools.deps and tools.deps.cli**
* 1.11.3.1452 on Apr 24, 2024
  * Update to Clojure 1.11.3 as default
* **1.11.2.1446 on Mar 8, 2024**
  * **Update to tools.deps 0.19.1417**
* **1.11.2.1441 on Mar 8, 2024**
  * **Update to tools.tools 0.3.3**
  * **Update to Clojure 1.11.2**
* **1.11.1.1435 on Dec 29, 2023**
  * **Update to tools.deps 1.18.1394**
* **1.11.1.1429 on Dec 4, 2023**
  * **Update to tools.deps 1.18.1374**
* **1.11.1.1420 on Dec 4, 2023**
  * **TDEPS-119 Unable to start CLI in write-protected project directory**
  * **Update to tools.deps 1.18.1370**
  * **Update to tools.tools 0.3.2**
* **1.11.1.1413 on Aug 26, 2023**
  * **No changes from prior**
* **1.11.1.1409 on Aug 25, 2023**
  * **No changes from prior**
* **1.11.1.1405 on Aug 22, 2023**
  * **GitHub releases now contain .sha256 files for binaries**
* **1.11.1.1386 on Aug 14, 2023**
  * **Release artifacts are now published to GitHub, and the download archive is used only as a mirror**
* **1.11.1.1369 on Aug 14, 2023**
  * **Download and mirroring wip**
* **1.11.1.1365 on Aug 13, 2023**
  * **Download and mirroring wip**
* **1.11.1.1347 on May 31, 2023**
  * **With -X or -T accept `-` as a trailing argument to read the remainder of args from stdin**
  * **On windows installer, hide progress bar on download**
  * **Update to tools.deps 0.18.1354**
  * **Update to tools.tools 0.3.1**
* **1.11.1.1273 on Apr 3, 2023**
  * **Change help text to point to -X:deps mvn-pom instead of -Spom**
  * **Switch to tools.deps 0.18.1335**
* **1.11.1.1267 on Mar 31, 2023**
  * **Switch to tools.deps 0.18.1331**
* **1.11.1.1262 on Mar 27, 2023**
  * **Switch to tools.deps 0.18.1317**
* **1.11.1.1257 on Mar 15, 2023**
  * **Switch to tools.deps 0.18.1308**
* **1.11.1.1252 on Mar 6, 2023**
  * **Switch to passing exec args via the basis :argmap**
  * **Function execution protocol support for -X/-T**
  * **Switch to tools.deps 0.17.1297**
* **1.11.1.1237 on Feb 27, 2023**
  * **Updates on -R and -C error messages**
* 1.11.1.1234 on Feb 27, 2023
  * Remove deprecated support for -R and -C
  * Clean up help text around repl supporting init-opts
  * Switch to tools.deps 0.16.1285
* **1.11.1.1224** on Feb 12, 2023**
  * **Switch to tools.deps 0.16.1281**
  * **TDEPS-236 Add rlwrap -m to clj for multiline editing**
* **1.11.1.1208 on Dec 11, 2022**
  * **Switch to tools.deps 0.16.1264**
  * **TDEPS-234 - Set -XX:-OmitStackTraceInFastThrow always**
* **1.11.1.1200 on Nov 14, 2022**
  * **Update to tools.deps.alpha 0.15.1254**
  * **Update to tools.tools v0.2.9**
* 1.11.1.1194 on Nov 14, 2022
  * Update to tools.deps.alpha 0.15.1250
* **1.11.1.1189 on Nov 2, 2022**
  * **TDEPS-233 bash script fixes from stale jar check changes**
  * **Add some missing items on help and man page**
* **1.11.1.1182 on Oct 24, 2022**
  * **Update to tools.deps.alpha 0.15.1244**
* 1.11.1.1177 on Oct 23, 2022
  * Fix bug in posix-install.sh
* 1.11.1.1174 on Oct 23, 2022
  * TDEPS-70 - Detect missing jar in classpath and recompute
  * TDEPS-200 - Clean up default user deps.edn
  * TDEPS-232 - Add more generic posix installer
  * Update to tools.deps.alpha 0.15.1237
* **1.11.1.1165 on Sep 18, 2022**
  * Extended the env var support to Windows
* 1.11.1.1161 on Sep 16, 2022
  * Use `$CLJ_JVM_OPTS` as jvm options on calls to internal tooling (making classpaths etc)
  * Use `$JAVA_OPTS` as jvm options on calls to user programs
* **1.11.1.1155 on Aug 5, 2022**
  * **Update to tools.deps.alpha 0.14.1222**
* **1.11.1.1149 on Jun 21, 2022**
  * **Update to tools.tools 0.2.8**
* 1.11.1.1145 on Jun 20, 2022
  * Update to tools.tools 0.2.7
  * Update to tools.deps.alpha 0.14.1212
* **1.11.1.1139 on Jun 16, 2022**
  * **Update to tools.tools 0.2.6**
  * **Update to tools.deps.alpha 0.14.1205**
* 1.11.1.1135 on Jun 16, 2022
  * Update to tools.deps.alpha 0.14.1200
* **1.11.1.1129 on Jun 12, 2022**
  * **Update to tools.deps.alpha 0.14.1194**
* **1.11.1.1124** on Jun 11, 2022

src/main/resources/clojure.ps1
#!/usr/bin/env pwsh

function Get-StringHash($str) {
  $md5 = new-Object -TypeName System.Security.Cryptography.MD5CryptoServiceProvider
  $utf8 = new-object -TypeName System.Text.UTF8Encoding
  return [System.BitConverter]::ToString($md5.ComputeHash($utf8.GetBytes($str)))
}

function IsNewerFile($file1, $file2) {
  if (!(Test-Path $file1)) {
    return $FALSE
  }
  if (!(Test-Path $file2)) {
    return $TRUE
  }
  $mod1 = (Get-ChildItem $file1).LastWriteTimeUtc
  $mod2 = (Get-ChildItem $file2).LastWriteTimeUtc
  return $mod1 -gt $mod2
}

$InstallDir="__PREFIX__"
$Version="__VERSION__"
$ToolsCp="$InstallDir/clojure-tools-$Version.jar"

$ErrorActionPreference = "Stop"

$PrintClassPath=$FALSE
$Describe=$FALSE
$Verbose=$FALSE
$Force=$FALSE
$Repro=$FALSE
$Tree=$FALSE
$Pom=$FALSE
$ResolveTags=$FALSE
$Help=$FALSE
$JvmOpts=@()
$ResolveAliases=@()
$ClasspathAliases=@()
$JvmAliases=@()
$MainAliases=@()
$AllAliases=@()

$params = $args
while ($params.Count -gt 0) {
  $arg, $params = $params
  if ($arg.StartsWith("-J")) {
    $JvmOpts += $arg.Substring(2)
  } elseif ($arg.StartsWith("-R")) {
    $aliases, $params = $params
    if ($aliases) {
      $ResolveAliases += $aliases
    } else {
      echo "Missing aliases"
      exit 1
    }
  } elseif ($arg.StartsWith("-C")) {
    $aliases, $params = $params
    if ($aliases) {
      $ClassPathAliases = += $aliases
    } else {
      echo "Missing aliases"
      exit 1
    }
  } elseif ($arg.StartsWith("-O")) {
    $aliases, $params = $params
    if ($aliases) {
      $JvmAliases = += $aliases
    } else {
      echo "Missing aliases"
      exit 1
    }
  } elseif ($arg.StartsWith("-M")) {
    $aliases, $params = $params
    if ($aliases) {
      $MainAliases = += $aliases
    } else {
      echo "Missing aliases"
      exit 1
    }
  } elseif ($arg.StartsWith("-A")) {
    $aliases, $params = $params
    if ($aliases) {
      $AllAliases = += $aliases
    } else {
      echo "Missing aliases"
      exit 1
    }
  } elseif ($arg -eq "-Sdeps") {
    $DepsData, $params = $params
    if (!($DepsData)) {
      echo "Missing deps"
      exit 1
    }
  } elseif ($arg -eq "-Scp") {
    $ForceCP, $params = $params
    if (!($ForceCP)) {
      echo "Missing path"
      exit 1
    }
  } elseif ($arg -eq "-Spath") {
    $PrintClassPath=$TRUE
  } elseif ($arg -eq "-Sverbose") {
    $Verbose=$TRUE
  } elseif ($arg -eq "-Sdescribe") {
    $Describe=$TRUE
  } elseif ($arg -eq "-Sforce") {
    $Force=$TRUE
  } elseif ($arg -eq "-Srepro") {
    $Repro=$TRUE
  } elseif ($arg -eq "-Stree") {
    $Tree=$TRUE
  } elseif ($arg -eq "-Spom") {
    $Pom=$TRUE
  } elseif ($arg -eq "-Sresolve-tags") {
    $ResolveTags=$TRUE
  } elseif ($arg.StartsWith("-S")) {
    echo "Invalid option: $arg"
    exit 1
  } elseif (($arg -eq "-h") -or ($arg -eq "--help") -or ($arg -eq "-?")) {
    if ($MainAliases.count -gt 0 -or $AllAliases.count -gt 0) {
      $ClojureArgs += @($arg) + $params
      break
    } else {
      $Help=$TRUE
    }
  } else {
    $ClojureArgs += @($arg) + $params
    break
  }
}

$JavaCmd=(Get-Command java.exe -ErrorAction SilentlyContinue).Path
if (!$JavaCmd) {
  if ($env:JAVA_HOME -and (Test-Path $env:JAVA_HOME )) {
    $JavaCmd="$JAVA_HOME/bin/java.exe"
  } else {
    echo "Couldn't find 'java'. Please set JAVA_HOME."
    exit 1
  }
}

if ($Help) {
  echo "
    Usage: clojure [dep-opt*] [init-opt*] [main-opt] [arg*] 
           clj     [dep-opt*] [init-opt*] [main-opt] [arg*] 

    The clojure script is a runner for Clojure. clj is a wrapper
    for interactive repl use. These scripts ultimately construct and
    invoke a command-line of the form:

    java [java-opt*] -cp classpath clojure.main [init-opt*] [main-opt] [arg*]

    The dep-opts are used to build the java-opts and classpath:
     -Jopt          Pass opt through in java_opts, ex: -J-Xmx512m
     -Oalias...     Concatenated jvm option aliases, ex: -O:mem
     -Ralias...     Concatenated resolve-deps aliases, ex: -R:bench:1.9
     -Calias...     Concatenated make-classpath aliases, ex: -C:dev
     -Malias...     Concatenated main option aliases, ex: -M:test
     -Aalias...     Concatenated aliases of any kind, ex: -A:dev:mem
     -Sdeps EDN     Deps data to use as the final deps file
     -Spath         Compute classpath and echo to stdout only
     -Scp CP        Do NOT compute or cache classpath, use this one instead
     -Srepro        Use only the local deps.edn (ignore other config files)
     -Sforce        Force recomputation of the classpath (don't use the cache)
     -Spom          Generate (or update existing) pom.xml with deps and paths
     -Stree         Print dependency tree
     -Sresolve-tags Resolve git coordinate tags to shas and update deps.edn
     -Sverbose      Print important path info to console
     -Sdescribe     Print environment and command parsing info as data

    init-opt:
     -i, --init path     Load a file or resource
     -e, --eval string   Eval exprs in string; print non-nil values

    main-opt:
     -m, --main ns-name  Call the -main function from namespace w/args
     -r, --repl          Run a repl
     path                Run a script from a file or resource
     -                   Run a script from standard input
     -h, -?, --help      Print this help message and exit

    For more info, see:
     https://clojure.org/guides/deps_and_cli
     https://clojure.org/reference/repl_and_main
"
  exit 0
}

if (!(Test-Path $InstallDir)) {
  echo "Clojure not installed."
  exit 1
} else {
  $ToolsCP = "$InstallDir/clojure-tools-${project.version}.jar"
}

if ($ResolveTags) {
  if (Test-Path deps.edn) {
    & "$JavaCmd" -classpath "$ToolsCP" clojure.main -m clojure.tools.deps.alpha.script.resolve-tags "--deps-file=deps.edn"
    exit 0
  } else {
    echo "deps.edn does not exist"
    exit 1
  }
}

$ConfigDir="$HOME/.clojure"
if (!(Test-Path "$ConfigDir")) {
  New-Item -Type Directory "$ConfigDir" | Out-Null
}

if (!(Test-Path "$ConfigDir/deps.edn")) {
  Copy-Item "$InstallDir/example-deps.edn" "$ConfigDir/deps.edn"
}

$UserCacheDir="$ConfigDir/.cpcache"
if ($Repro) {
  $ConfigPaths=@("$InstallDir/deps.edn", "deps.edn")
} else {
  $ConfigPaths=@("$InstallDir/deps.edn", "$ConfigDir/deps.edn", "deps.edn")
}
$ConfigStr=[String]::Join(",", $ConfigPaths)

if (Test-Path "deps.edn") {
  $CacheDir=".cpcache"
} else {
  $CacheDir="$UserCacheDir"
}

$CK="$ResolveAliases|$ClassPathAliases|$AllAliases|$JvmAliases|$MainAliases|" + "$DepsData|" + [String]::Join("|", $ConfigPaths)
$CK=(Get-StringHash $CK) | %{$_ -replace "-", ""}

$LibsFile="$CacheDir/$CK.libs"
$CpFile="$CacheDir/$CK.cp"

if ($Verbose) {
  echo "version      = ${project.version}"
  echo "install_dir  = $InstallDir"
  echo "config_dir   = $ConfigDir"
  echo "config_paths = $ConfigPaths"
  echo "cache_dir    = $CacheDir"
  echo "cp_file      = $CpFile"
  echo ""
}

$Stale=$FALSE
if ($Force -or !(Test-Path "$CpFile")) {
  $Stale=$TRUE
} else {
  foreach ($ConfigPath in $ConfigPaths) {
    if (IsNewerFile $ConfigPath $CpFile) {
      $Stale=$TRUE
      break
    }
  }
}

if ($Stale -or $Pom) {
  $ToolsArgs=@()
  if ($DepsData) {
    $ToolsArgs += "--config-data"
    $ToolsArgs += $DepsData
  }
  if ($ResolveAliases) {
    $ToolsArgs += "-R$ResolveAliases"
  }
  if ($ClassPathAliases) {
    $ToolsArgs += "-C$ClassPathAliases"
  }
  if ($JvmAliases) {
    $ToolsArgs += "-J$JvmAliases"
  }
  if ($MainAliases) {
    $ToolsArgs += "-M$MainAliases"
  }
  if ($AllAliases) {
    $ToolsArgs += "-A$AllAliases"
  }
  if ($ForceCp) {
    $ToolsArgs += "--skip-cp"
  }
}

if ($Stale) {
  if ($Verbose) {
    echo "Refreshing classpath"
  }

  & "$JavaCmd" -Xmx256m -classpath "$ToolsCp" clojure.main -m clojure.tools.deps.alpha.script.make-classpath "--config-files=$ConfigStr" "--libs-file=$LibsFile" "--cp-file=$CpFile" "--jvm-file=$JvmFile" "--main-file=$MainFile" @ToolsArgs
}

if ($ForceCp) {
  $CP=$ForceCp
} else {
  $CP=(Get-Content "$CpFile")
}

if ($Pom) {
  & "$JavaCmd" -Xmx256m -classpath "$ToolsCp" clojure.main -m clojure.tools.deps.alpha.script.generate-manifest "--config-files=$ConfigStr" --gen=pom @ToolsArgs
} elseif ($PrintClassPath) {
  echo "$CP"
} elseif ($Describe) {
  foreach ($ConfigPath in $ConfigPaths) {
    if (Test-Path "$ConfigPath") {
      $PathVector = "$PathVector`"$ConfigPath`" "
    }
  }
  echo "
    {:version \"${project.version}\"
     :config-files [$PathVector]
     :install-dir $InstallDir
     :config-dir $ConfigDir
     :cache-dir $CacheDir
     :force $Force
     :repro $Repro
     :resolve-aliases "$(join '' ${resolve_aliases[@]})"
     :classpath-aliases "$(join '' ${classpath_aliases[@]})"
     :jvm-aliases "$(join '' ${jvm_aliases[@]})"
     :main-aliases "$(join '' ${main_aliases[@]})"
     :all-aliases "$(join '' ${all_aliases[@]})"}
  "
} elseif ($Tree) {
  & "$JavaCmd" -Xmx256m -classpath "$ToolsCp" clojure.main -m clojure.tools.deps.alpha.script.print-tree "--libs-file=$LibsFile"
} else {
  # TODO handle jvm and main cache opts
  & "$JavaCmd" @JvmOpts -classpath "$CP" "-Dclojure.libfile=$LibsFile" clojure.main @ClojureArgs
}